sklearn/preprocessing/tests/test_knn_imputation.py

```diff
@@ -0,0 +1,432 @@
+    ])
+    assert_raises(ValueError, imputer.fit, X)
+
+    # Test with an imputable matrix and also compare with missing_values="NaN"
+    # X = np.array([
+    #     [1, 0, 1, 0, 5],
+    #     [2, 1, 2, 2, 3],
+    #     [3, 2, 3, 0, 0],
+    #     [6, 6, 0, 5, 13],
+    # ])
+
+    X = np.array([
+        [1, 0, 1, 0, 1],
+        [2, 1, 2, 2, 3],
+        [3, 2, 3, 0, 0],
+        [6, 6, 0, 5, 17],
+    ])
+
+    # X_nan = np.array([
+    #     [1, np.nan, 1, np.nan, 1],
+    #     [2, 1, 2, 2, 3],
+    #     [3, 2, 3, np.nan, np.nan],
+    #     [6, 6, np.nan, 5, 17],
+    # ])
+
+    statistics_mean = [3, 3, 2, 3.5, 7]
+    X_imputed = np.array([
+        [1, 1.5, 1,   2, 1],
+        [2, 1,   2,   2, 3],
+        [3, 2,   3,   2, 2],
+        [6, 6,   1.5, 5, 17],
+    ])
+
+    assert_array_equal(imputer.fit_transform(X), X_imputed)
+    assert_array_equal(imputer.statistics_, statistics_mean)
+    # The following fails at the moment as NearestNeighbors object does not
+    # pass missing_values=0 to pairwise_distances()
+    # assert_array_equal(imputer.fit_transform(X), imputer_nan.fit_transform(
+    #     X_nan))
+
+
+def test_knn_imputation_default():
+    # Test imputation with default values
+    # imputer = KNNImputer()
+
+    # Test with % missing in a column > col_max_missing
+    X = np.array([
+        [np.nan, 0, 0, 0, 5],
+        [np.nan, 1, 0, np.nan, 3],
+        [np.nan, 2, 0, 0, 0],
+        [np.nan, 6, 0, 5, 13],
+        [np.nan, 7, 0, 7, 8],
+        [np.nan, 8, 0, 8, 9],
+    ])
+    assert_raises(ValueError, KNNImputer().fit, X)
+
+    # Test with insufficient number of neighbors
+    imputer = KNNImputer()
+    X = np.array([
+        [1, 1, 1, 2, np.nan],
+        [2, 1, 2, 2, 3],
+        [3, 2, 3, 3, 8],
+        [6, 6, 2, 5, 13],
+    ])
+    assert_raises(ValueError, KNNImputer().fit, X)
+
+    # Test with inf present
+    X = np.array([
+        [np.inf, 1, 1, 2, np.nan],
+        [2, 1, 2, 2, 3],
+        [3, 2, 3, 3, 8],
+        [np.nan, 6, 0, 5, 13],
+        [np.nan, 7, 0, 7, 8],
+        [6, 6, 2, 5, 7],
+    ])
+    assert_raises(ValueError, KNNImputer().fit, X)
+
+    # Test with an imputable matrix
+    X = np.array([
+        [1,      0,      0,      1],
+        [2,      1,      2,      np.nan],
+        [3,      2,      3,      np.nan],
+        [np.nan, 4,      5,      5],
+        [6,      np.nan, 6,      7],
+        [8,      8,      8,      8],
+        [16,     15,     18,    19],
+    ])
+
+    statistics_mean = [6, 5, 6, 8]
+
+    X_imputed = np.array([
+        [1,      0,      0,      1],
+        [2,      1,      2,      5.25],
+        [3,      2,      3,      5.25],
+        [4,      4,      5,      5],
+        [6,      3,      6,      7],
+        [8,      8,      8,      8],
+        [16,     15,     18,    19],
+    ])
+
+    imputer = KNNImputer()
+    assert_array_equal(imputer.fit_transform(X), X_imputed)
+    assert_array_equal(imputer.statistics_, statistics_mean)
```

Comment:
instead of testing against something hard coded, test against what Imputer produces, perhaps?